



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="A package with utilities for next generation sequencing">
      
      
      
        <meta name="author" content="Ben Murrell">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0, mkdocs-material-3.0.3">
    
    
      
        <title>hmm - NextGenSeqUtils.jl</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.451f80e5.css">
      
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../assets/Documenter.css">
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="../#hidden-markov-model-functions" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href=".." title="NextGenSeqUtils.jl" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                NextGenSeqUtils.jl
              </span>
              <span class="md-header-nav__topic">
                hmm
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/MurrellGroup/NextGenSeqUtils.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href=".." title="NextGenSeqUtils.jl" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    NextGenSeqUtils.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/MurrellGroup/NextGenSeqUtils.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../align/" title="Align" class="md-nav__link">
      Align
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
    <a href="./" title="hmm" class="md-nav__link md-nav__link--active">
      hmm
    </a>
    
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../io/" title="io" class="md-nav__link">
      io
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../kmers/" title="kmers" class="md-nav__link">
      kmers
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../orient/" title="orient" class="md-nav__link">
      orient
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../paths/" title="paths" class="md-nav__link">
      paths
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../phreds/" title="phreds" class="md-nav__link">
      phreds
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../simulation/" title="simulation" class="md-nav__link">
      simulation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../utils/" title="utils" class="md-nav__link">
      utils
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/MurrellGroup/NextGenSeqUtils.jl/edit/master/docs/hmm.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='Hidden-Markov-Model-Functions-1'></a></p>
<h1 id="hidden-markov-model-functions">Hidden Markov Model Functions</h1>
<ul>
<li><a href="./#NextGenSeqUtils.backward_logs"><code>NextGenSeqUtils.backward_logs</code></a></li>
<li><a href="./#NextGenSeqUtils.forward_backward_logs"><code>NextGenSeqUtils.forward_backward_logs</code></a></li>
<li><a href="./#NextGenSeqUtils.forward_logs"><code>NextGenSeqUtils.forward_logs</code></a></li>
<li><a href="./#NextGenSeqUtils.gen_seq_with_model"><code>NextGenSeqUtils.gen_seq_with_model</code></a></li>
<li><a href="./#NextGenSeqUtils.get_obs_given_state"><code>NextGenSeqUtils.get_obs_given_state</code></a></li>
<li><a href="./#NextGenSeqUtils.homopolymer_filter"><code>NextGenSeqUtils.homopolymer_filter</code></a></li>
<li><a href="./#NextGenSeqUtils.initial_dist"><code>NextGenSeqUtils.initial_dist</code></a></li>
<li><a href="./#NextGenSeqUtils.logsum"><code>NextGenSeqUtils.logsum</code></a></li>
<li><a href="./#NextGenSeqUtils.markov_filter"><code>NextGenSeqUtils.markov_filter</code></a></li>
<li><a href="./#NextGenSeqUtils.obs_mat"><code>NextGenSeqUtils.obs_mat</code></a></li>
<li><a href="./#NextGenSeqUtils.trans_mat"><code>NextGenSeqUtils.trans_mat</code></a></li>
<li><a href="./#NextGenSeqUtils.viterbi_logs"><code>NextGenSeqUtils.viterbi_logs</code></a></li>
<li><a href="./#NextGenSeqUtils.viterbiprint"><code>NextGenSeqUtils.viterbiprint</code></a></li>
</ul>
<p><a id='NextGenSeqUtils.viterbi_logs' href='#NextGenSeqUtils.viterbi_logs'>#</a>
<strong><code>NextGenSeqUtils.viterbi_logs</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>viterbi_logs(observations_given_states::Array{Float64, 2}, transitions::Array{Float64, 2}, initials::Array{Float64})
</code></pre>

<p>Return viterbi path and log probability for that path. Takes logs of matrices. <code>observations_given_states</code> has # rows = # states, # columns = # steps of markov process.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L11-L16' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.trans_mat' href='#NextGenSeqUtils.trans_mat'>#</a>
<strong><code>NextGenSeqUtils.trans_mat</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>trans_mat(; uniform_cycle_prob = 0.9999999999, homopoly_cycle_prob = 0.98)
</code></pre>

<p>Return 5x5 transition matrix with given transition probabilities State 1 – uniform observation distribution. State 2 – High "A" observation likelihood. State 3 – High "C" observation likelihood. State 4 – High "G" observation likelihood. State 5 – High "T" observation likelihood. Each state has high likelihood to return to state 1. States 2 through 5 represent high likelihood of homopolymer regions of respective nuc.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L42-L52' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.obs_mat' href='#NextGenSeqUtils.obs_mat'>#</a>
<strong><code>NextGenSeqUtils.obs_mat</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>obs_mat(; homopoly_prob = 0.99)
</code></pre>

<p>Return 5x4 observation matrix with given probabilities. State 1 – uniform observation distribution. State 2 – High "A" observation likelihood. State 3 – High "C" observation likelihood. State 4 – High "G" observation likelihood. State 5 – High "T" observation likelihood. Columns are in same order as columns 2 through 5 (the last four rows give a symmetric matrix).</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L70-L81' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.initial_dist' href='#NextGenSeqUtils.initial_dist'>#</a>
<strong><code>NextGenSeqUtils.initial_dist</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>initial_dist(; uniform_state = 0.99)
</code></pre>

<p>Initial state distributions: see <code>trans_mat</code> for state descriptions. 5x1 vector.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L99-L103' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.get_obs_given_state' href='#NextGenSeqUtils.get_obs_given_state'>#</a>
<strong><code>NextGenSeqUtils.get_obs_given_state</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>get_obs_given_state(observation_matrix::Array{Float64,2}, observation_seq::String)
</code></pre>

<p>Populate 5xT matrix with likelihood of observation at each of T time steps given each  nucleotide in <code>observation_seq</code>.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L109-L114' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.homopolymer_filter' href='#NextGenSeqUtils.homopolymer_filter'>#</a>
<strong><code>NextGenSeqUtils.homopolymer_filter</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>homopolymer_filter(seqs::Array{String,1}, phreds, names;
                   transmat = nothing, obsmat = nothing,
                   initialdist = nothing)
</code></pre>

<p>Filter sequences with "bad" sections in the middle – abnormally long runs of a single base, using viterbi alg inference. If this homopolymer occurs on one end of a sequence, keeps sequence and trims homopolymer region off end. <code>phreds</code> and/or <code>names</code> may be <code>nothing</code>, in which case <code>nothing</code> is returned for the respective field. If transition, observation, initial distribution matrices not provided (default) then the default values from the respective constructors are used.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L126-L136' class='documenter-source'>source</a><br></p>
<pre><code>homopolymer_filter(sourcepath::String, destpath::String;
                   transmat = nothing, obsmat = nothing,
                   initialdist = nothing, format=&quot;fastq&quot;)
</code></pre>

<p>Filter sequences with "bad" sections in the middle – abnormally long runs of a single base, using viterbi alg inference. If this homopolymer occurs on one end of a sequence, keeps sequence and trims homopolymer region off end. Takes a file path for each of a source file of type <code>format</code> (which may be "fasta" or "fastq") and a destination file is written which is the same file type.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L191-L200' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.markov_filter' href='#NextGenSeqUtils.markov_filter'>#</a>
<strong><code>NextGenSeqUtils.markov_filter</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>homopolymer_filter(seqs::Array{String,1}, phreds, names;
                   transmat = nothing, obsmat = nothing,
                   initialdist = nothing)
</code></pre>

<p>Filter sequences with "bad" sections in the middle – abnormally long runs of a single base, using viterbi alg inference. If this homopolymer occurs on one end of a sequence, keeps sequence and trims homopolymer region off end. <code>phreds</code> and/or <code>names</code> may be <code>nothing</code>, in which case <code>nothing</code> is returned for the respective field. If transition, observation, initial distribution matrices not provided (default) then the default values from the respective constructors are used.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L126-L136' class='documenter-source'>source</a><br></p>
<pre><code>homopolymer_filter(sourcepath::String, destpath::String;
                   transmat = nothing, obsmat = nothing,
                   initialdist = nothing, format=&quot;fastq&quot;)
</code></pre>

<p>Filter sequences with "bad" sections in the middle – abnormally long runs of a single base, using viterbi alg inference. If this homopolymer occurs on one end of a sequence, keeps sequence and trims homopolymer region off end. Takes a file path for each of a source file of type <code>format</code> (which may be "fasta" or "fastq") and a destination file is written which is the same file type.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L191-L200' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.forward_logs' href='#NextGenSeqUtils.forward_logs'>#</a>
<strong><code>NextGenSeqUtils.forward_logs</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>forward_logs(observations_given_states::Array{Float64, 2}, transitions::Array{Float64, 2}, initials::Array{Float64})
</code></pre>

<p>Compute logs of forward scores. Takes logs of matrices as inputs. <code>observations_given_states</code> has # rows = # states, # columns = # steps of markov process.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L224-L229' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.backward_logs' href='#NextGenSeqUtils.backward_logs'>#</a>
<strong><code>NextGenSeqUtils.backward_logs</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>backward_logs(observations_given_states::Array{Float64, 2}, transitions::Array{Float64, 2}, initials::Array{Float64})
</code></pre>

<p>Compute logs of backward scores and individual posterior probabilities. Takes logs of matrices as inputs. <code>observations_given_states</code> has # rows = # states, # columns = # steps of markov process.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L244-L249' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.forward_backward_logs' href='#NextGenSeqUtils.forward_backward_logs'>#</a>
<strong><code>NextGenSeqUtils.forward_backward_logs</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>forward_backward_logs(observations_given_states::Array{Float64, 2}, transitions::Array{Float64, 2}, initials::Array{Float64})
</code></pre>

<p>Compute logs of forward-backward scores and individual probabilities. Takes logs of matrices as inputs. <code>observations_given_states</code> has # rows = # states, # columns = # steps of markov process.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L264-L269' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.logsum' href='#NextGenSeqUtils.logsum'>#</a>
<strong><code>NextGenSeqUtils.logsum</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>logsum(lga, lgb)
</code></pre>

<p>Compute numerically stable logsum. Returns <code>-Inf</code> if either input is <code>-Inf</code>.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/utils.jl#L348-L352' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.gen_seq_with_model' href='#NextGenSeqUtils.gen_seq_with_model'>#</a>
<strong><code>NextGenSeqUtils.gen_seq_with_model</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>gen_seq_with_model(n::Int, trans_mat, obs_mat, initial_dists)
</code></pre>

<p>Generate a sequence given a markov model. May create bad reads (long runs of a single base).</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L294-L298' class='documenter-source'>source</a><br></p>
<p><a id='NextGenSeqUtils.viterbiprint' href='#NextGenSeqUtils.viterbiprint'>#</a>
<strong><code>NextGenSeqUtils.viterbiprint</code></strong> &mdash; <em>Function</em>.</p>
<pre><code>viterbiprint(s::String)
</code></pre>

<p>Draw flagged sites with capital letters, safe sites with lowercase.</p>
<p><a target='_blank' href='https://github.com/MurrellGroup/NextGenSeqUtils.jl/blob/fdf03a6fa43baf30257d70c58d0c345ad0d25065/src/hmm.jl#L310-L314' class='documenter-source'>source</a><br></p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../align/" title="Align" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Align
              </span>
            </div>
          </a>
        
        
          <a href="../io/" title="io" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                io
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.583bbe55.js"></script>
      
      <script>app.initialize({version:"1.0",url:{base:".."}})</script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_HTML"></script>
      
        <script src="../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>